src/pages/posts/comparing-rust-smart-contract-storage-approaches.md

```diff
@@ -0,0 +1,515 @@
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Increment {} => {
+            STATE.update(deps.storage, |mut state| -> Result<_, ContractError> {
+                state.count += 1;
+                Ok(state)
+            })?;
+
+            Ok(Response::new())
+        }
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetNum {} => {
+            let state = STATE.load(deps.storage)?;
+            let raw_response = GetNumResponse { count: state.count };
+            to_binary(&raw_response)
+        }
+    }
+}
+```
+
+Phew! Still with me? Let's step through it.
+
+### Imports
+
+Casual.
+
+```rust
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{
+    entry_point, to_binary, Binary, Deps, DepsMut, Env, MessageInfo, Response, StdError, StdResult,
+};
+use cw_storage_plus::Item;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+```
+
+Something interesting is just the number of dependencies. 
+
+1. `cosmwasm_schema`
+2. `cosmwasm_std`
+3. `cw_storage_plus`
+4. `schemars`
+5. `serde`
+6. `thiserror`
+
+You might think this results in large contract file sizes, but with the [CosmWasm optimizer](https://github.com/CosmWasm/rust-optimizer), it's actually not too far out of [what I've come to expect](https://github.com/stellar/sorobanathon/discussions/26) with NEAR contracts. Somewhere in the 150kB range.
+
+And that optimizer gives you something the NEAR ecosystem sorely lacks: **reproducible builds!** The same contract results in the same optimized Wasm, no matter who builds it when. Which leads us right into the next interesting thing:
+
+### Initialization
+
+Unlike NEAR and Soroban, every CosmWasm contract needs to be initialized before it can be called. Here's all the stuff from above that's just there to deal with initialization:
+
+```rust
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub count: i8,
+}
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let state = State { count: msg.count };
+    STATE.save(deps.storage, &state)?;
+    Ok(Response::new())
+}
+```
+
+You probably noticed this `cw_serde` macro on all the structs and enums. It works together with a [`cargo schema` alias](https://github.com/CosmWasm/cw-template/blob/main/.cargo/config) and [accompanying embedded package](https://github.com/CosmWasm/cw-template/blob/231956d54770137fc0c5147f0128c9afd572eaf1/src/bin/schema.rs#L1) to generate JSON schemas for all your contract calls. These work like Ethereum's ABIs (or, you know, [like RAEN](https://youtu.be/VenoNgWdvw0) for NEAR 😏).
+
+Anyhow, the interesting part is that it **needs** to be initialized. Why?
+
+Because a given set of Wasm bytes is **only deployed once**, in CosmWasm.
+
+Let me say that again; it's super cool.
```

Comment:
Cool indeed! So cool that we're also doing this [in Soroban](https://github.com/stellar/rs-soroban-env/pull/572)  :)  

Hopefully to be released tomorrow and will hit the sdk and examples by end-of-week

Reply:
Very exciting!